File: laba3/task1/Golang/hashchain.go
package main
import "fmt"

type NodeHash struct {
	key  string
	data int
	next *NodeHash
	prev *NodeHash
}

func NewNodeHash(k string, d int) *NodeHash {
	return &NodeHash{key: k, data: d, next: nil, prev: nil}
}

func (this *NodeHash) getKey() string     { return this.key }
func (this *NodeHash) getData() int       { return this.data }
func (this *NodeHash) getNext() *NodeHash { return this.next }

type Hash struct {
	cell     []*NodeHash
	capacity int
	size     int
}

func NewHash() *Hash {
	return &Hash{
		cell:     make([]*NodeHash, 10),
		capacity: 10,
		size:     0,
	}
}

func (this *Hash) hashFunc(key string, capacity int) int {
	hash := 0
	for _, c := range key {
		hash = (hash*37 + int(c)) % capacity
	}
	return hash % capacity
}

func (this *Hash) addElement(key string, data int) {
	index := this.hashFunc(key, this.capacity)
	begin := this.cell[index]
	current := begin
	for current != nil {
		if current.key == key {
			current.data = data
			return
		}
		current = current.next
	}
	node := NewNodeHash(key, data)
	node.next = begin
	node.prev = nil
	if this.cell[index] != nil {
		this.cell[index].prev = node
	}
	this.cell[index] = node
	this.size++
}

func (this *Hash) findElement(key string) int {
	index := this.hashFunc(key, this.capacity)
	current := this.cell[index]
	for current != nil {
		if current.key == key {
			return current.data
		}
		current = current.next
	}
	return -1
}

func (this *Hash) deleteElement(key string) {
	index := this.hashFunc(key, this.capacity)
	delVal := this.cell[index]
	for delVal != nil {
		if delVal.key == key {
			if delVal.prev != nil {
				delVal.prev.next = delVal.next
			} else {
				this.cell[index] = delVal.next
			}
			if delVal.next != nil {
				delVal.next.prev = delVal.prev
			}
			this.size--
			return
		}
		delVal = delVal.next
	}
}

func (this *Hash) printHashTable() {
	fmt.Printf("Хэш-таблица (size: %d, capacity: %d):\n", this.size, this.capacity)
	for i := 0; i < this.capacity; i++ {
		fmt.Printf("Ячейка %d: ", i)
		current := this.cell[i]
		if current == nil {
			fmt.Print("empty")
		} else {
			for current != nil {
				fmt.Printf("[%s -> %d]", current.key, current.data)
				if current.next != nil {
					fmt.Print(" -> ")
				}
				current = current.next
			}
		}
		fmt.Println()
	}
}

func (this *Hash) getCapacity() int       { return this.capacity }
func (this *Hash) getSize() int           { return this.size }
func (this *Hash) getCells() []*NodeHash { return this.cell }

File: laba3/task3/C++/hashchain.cpp
#include <iostream>
#include <string>
#include "structures.h"

using namespace std;

int Hash::hashFunc(const string& key, const int capacity){
    int hash = 0;
    for (const char c: key) {
        hash = (hash * 37 + c) % capacity;
    }
    return hash % capacity;
}

void Hash::addElement(const string& key, const int data){
    const int index = hashFunc(key, this->capacity);
    NodeHash* begin = this->cell[index];
    NodeHash* current = begin;
    
    // Проверяем, существует ли уже ключ
    while (current != nullptr){
        if (current->getKey() == key){
            current->data = data; // Прямое обращение, так как мы friend
            return;
        }
        current = current->getNext();
    }
    
    // Создаем новый узел
    auto* node = new NodeHash(key, data);
    node->next = begin;
    node->prev = nullptr;
    
    if (begin != nullptr) {
        begin->prev = node;
    }
    
    this->cell[index] = node;
    this->size++;
}

int Hash::findElement(const string& key) const
{
    const int index = hashFunc(key, this->capacity);
    NodeHash* current = this->cell[index];
    
    while (current != nullptr){
        if (current->getKey() == key){
            return current->getData();
        }
        current = current->getNext();
    }
    return -1;
}

void Hash::deleteElement(const string& key){
    const int index = hashFunc(key, this->capacity);
    NodeHash* current = this->cell[index];
    
    while (current != nullptr){
        if (current->getKey() == key){
            // Обновляем указатели соседних узлов
            if (current->prev != nullptr) {
                current->prev->next = current->next;
            } else {
                this->cell[index] = current->next;
            }
            
            if (current->next != nullptr) {
                current->next->prev = current->prev;
            }
            
            delete current;
            this->size--;
            return;
        }
        current = current->getNext();
    }
}

void Hash::printHashTable() const
{
    cout << "Хэш-таблица (size: " << this->size << ", capacity: " << this->capacity << "):" << endl;
    
    for (int i = 0; i < this->capacity; i++) {
        cout << "Ячейка " << i << ": ";

        NodeHash* current = this->cell[i];
        if (current == nullptr) {
            cout << "empty";
        } else {
            bool first = true;
            while (current != nullptr) {
                if (!first) {
                    cout << " -> ";
                }
                cout << "[" << current->getKey() << " -> " << current->getData() << "]";
                current = current->getNext();
                first = false;
            }
        }
        cout << endl;
    }
}

Hash::Hash(const Hash& other) {
    this->capacity = other.capacity;
    this->size = 0;
    this->cell = new NodeHash*[this->capacity];
    
    for (int i = 0; i < this->capacity; i++) {
        this->cell[i] = nullptr;
    }
    
    for (int i = 0; i < other.capacity; i++) {
        NodeHash* current = other.cell[i];
        while (current != nullptr) {
            this->addElement(current->getKey(), current->getData());
            current = current->getNext();
        }
    }
}

Hash& Hash::operator=(const Hash& other) {
    if (this == &other) {
        return *this;
    }
    
    for (int i = 0; i < capacity; i++) {
        NodeHash* current = cell[i];
        while (current != nullptr) {
            NodeHash* next = current->getNext();
            delete current;
            current = next;
        }
    }
    delete[] cell;
    this->capacity = other.capacity;
    this->size = 0;
    this->cell = new NodeHash*[this->capacity];
    
    for (int i = 0; i < this->capacity; i++) {
        this->cell[i] = nullptr;
    }
    
    for (int i = 0; i < other.capacity; i++) {
        NodeHash* current = other.cell[i];
        while (current != nullptr) {
            this->addElement(current->getKey(), current->getData());
            current = current->getNext();
        }
    }
    
    return *this;
}
